tests/test_charm.py

```diff
@@ -88,12 +88,40 @@ def test_metrics_endpoint_relation(self, mock_remove_user, mock_add_user,
         harness.remove_relation(relation_id)
         mock_remove_user.assert_called_once_with(f'juju-metrics-r{relation_id}')
 
+    @patch("ops.model.Model.get_binding")
+    def test_dbcluster_relation_changed_single_addr(self, binding):
+        harness = self.harness
+        binding.return_value = mockBinding(["192.168.1.17"])
+
+        relation_id = harness.add_relation('dbcluster', 'controller')
+        harness.add_relation_unit(relation_id, 'juju-controller/1')
+
+        harness.charm._on_dbcluster_relation_changed(
+            harness.charm.model.get_relation('dbcluster').data[harness.charm.unit])
+
+        data = harness.get_relation_data(relation_id, 'juju-controller/0')
+        self.assertEqual(data["db-bind-address"], "192.168.1.17")
+
+    @patch("ops.model.Model.get_binding")
+    def test_dbcluster_relation_changed_multi_addr_error(self, binding):
+        harness = self.harness
+        binding.return_value = mockBinding(["192.168.1.17", "192.168.1.18"])
```

Comment:
personally I'd use a MagicMock instead of creating these custom types but that's a preference.

```
binding = MagicMock()
binding.network.ingress_addresses = [...]
binding.network.ingress_address = ...
```